File: RegExpBuilder.py
import re

class RegExpBuilder:

    def __init__(self):
        self._literal = ""
        self._ignoreCase = False
        self._multiLine = False
        self._specialCharactersInsideCharacterClass = set(["^", "-", "]"])
        self._specialCharactersOutsideCharacterClass = set([".", "^", "$", "*", "+", "?", "(", ")", "[", "{"])
        self._min = -1
        self._max = -1
        self._of = ""
        self._ofAny = False
        self._ofGroup = -1
        self._from = ""
        self._notFrom = ""
        self._like = ""
        self._either = ""
        self._reluctant = False
        self._capture = False

    def _clear(self):
        self._ignoreCase = False
        self._multiLine = False
        self._min = -1
        self._max = -1
        self._of = ""
        self._ofAny = False
        self._ofGroup = -1
        self._from = ""
        self._notFrom = ""
        self._like = ""
        self._either = ""
        self._reluctant = False
        self._capture = False

    def _flush_state(self):
        if self._of != "" or self._ofAny or self._ofGroup > 0 or self._from != "" or self._notFrom != "" or self._like != "":
            captureLiteral = "" if self._capture else "?:"
            quantityLiteral = self._get_quantity_literal()
            characterLiteral = self._get_character_literal()
            reluctantLiteral = "?" if self._reluctant else ""
            self._literal += "(" + captureLiteral + "(?:" + characterLiteral + ")" + quantityLiteral + reluctantLiteral + ")"
            self._clear()
  
    def _get_quantity_literal(self):
        if self._min != -1:
            if self._max != -1:
                return "{" + str(self._min) + "," + str(self._max) + "}"
            return "{" + str(self._min) + ",}"
        return "{0," + str(self._max) + "}"
  
    def _get_character_literal(self):
        if self._of != "":
            return self._of
        if self._ofAny:
            return "."
        if self._ofGroup > 0:
            return "\\" + str(self._ofGroup)
        if self._from != "":
            return "[" + self._from + "]"
        if self._notFrom != "":
            return "[^" + self._notFrom + "]"
        if self._like != "":
            return self._like
  
    def get_literal(self):
        self._flush_state()
        return self._literal
  
    def get_regexp(self):
        self._flush_state()
        flags = 0
        if self._ignoreCase:
            flags = flags | re.IGNORECASE
        if self._multiLine:
            flags = flags | re.MULTILINE
        return re.compile(self._literal, flags)
  
    def ignore_case(self):
        self._ignoreCase = True
        return self
  
    def multi_line(self):
        self._multiLine = True
        return self
  
    def start_of_input(self):
        self._literal += "(?:^)"
        return self

    def start_of_line(self):
        self.multi_line()
        return self.start_of_input()
  
    def end_of_input(self):
        self._flush_state()
        self._literal += "(?:$)"
        return self

    def end_of_line(self):
        self.multi_line()
        return self.end_of_input()

    def either(self, s):
        if len(s) > 0:
            self._flush_state()
            literal = "(?:"
            for element in s:
                literal += "(?:"
                if type(element) is str:
                    literal += RegExpBuilder().exactly(1).of(element).get_literal()
                else:
                    literal += element.get_literal()
                literal += ")|"
            literal = literal[:-1] + ")"
            self._literal += literal
        return self
  
    def exactly(self, n):
        self._flush_state()
        self._min = n
        self._max = n
        return self
  
    def min(self, n):
        self._flush_state()
        self._min = n
        return self
  
    def max(self, n):
        self._flush_state()
        self._max = n
        return self
  
    def of(self, s):
        self._of = self._escape_outside_character_class(s)
        return self
  
    def of_any(self):
        self._ofAny = True
        return self

    def of_group(self, n):
        self._ofGroup = n
        return self
  
    def from_class(self, s):
        self._from = self._escape_inside_character_class("".join(s))
        return self
  
    def not_from_class(self, s):
        self._notFrom = self._escape_inside_character_class("".join(s))
        return self
  
    def like(self, r):
        self._like = r.get_literal()
        return self
  
    def reluctantly(self):
        self._reluctant = True
        return self
  
    def ahead(self, r):
        self._flush_state()
        self._literal += "(?=" + r.get_literal() + ")"
        return self
  
    def not_ahead(self, r):
        self._flush_state()
        self._literal += "(?!" + r.get_literal() + ")"
        return self
  
    def as_group(self):
        self._capture = True
        return self

    def then(self, s):
        return self.exactly(1).of(s)

    def some(self, s):
        return self.min(1).from_class(s)

    def maybe_some(self, s):
        return self.min(0).from_class(s)

    def maybe(self, s):
        return self.max(1).of(s)

    def something(self):
        return self.min(1).of_any()

    def anything(self):
        return self.min(0).of_any()

    def any(self):
        return self.exactly(1).of_any()

    def line_break(self):
        return self.either(["\r\n", "\r", "\n"])

    def line_breaks(self):
        return self.like(RegExpBuilder().line_break())

    def whitespace(self):
        if self._min == -1 and self._max == -1:
            return self.exactly(1).of("\s")
        self._like = "\s"
        return self

    def not_whitespace(self):
        if self._min == -1 and self._max == -1:
            return self.exactly(1).of("\S")
        self._like = "\S"
        return self

    def tab(self):
        return self.exactly(1).of("\t")

    def tabs(self):
        return self.like(RegExpBuilder().tab())

    def digit(self):
        return self.exactly(1).of("\d")

    def not_digit(self):
        return self.exactly(1).of("\D")

    def digits(self):
        return self.like(RegExpBuilder().digit())

    def not_digits(self):
        return self.like(RegExpBuilder().not_digit())

    def letter(self):
        self.exactly(1)
        self._from = "A-Za-z"
        return self

    def not_letter(self):
        self.exactly(1)
        self._notFrom = "A-Za-z"
        return self

    def letters(self):
        self._from = "A-Za-z"
        return self

    def not_letters(self):
        self._not_from = "A-Za-z"
        return self

    def lower_case_letter(self):
        self.exactly(1)
        self._from = "a-z"
        return self

    def lower_case_letters(self):
        self._from = "a-z"
        return self

    def upper_case_letter(self):
        self.exactly(1)
        self._from = "A-Z"
        return self

    def upper_case_letters(self):
        self._from = "A-Z"
        return self

    def append(self, r):
        self.exactly(1)
        self._like = r.get_literal()
        return self

    def optional(self, r):
        self.max(1)
        self._like = r.get_literal()
        return self
  
    def _escape_inside_character_class(self, s):
        return self._escape_special_characters(s, self._specialCharactersInsideCharacterClass)

    def _escape_outside_character_class(self, s):
        return self._escape_special_characters(s, self._specialCharactersOutsideCharacterClass)
  
    def _escape_special_characters(self, s, specialCharacters):
        escapedString = ""
        for i in range(len(s)):
            character = s[i]
            if character in specialCharacters:
                escapedString += "\\" + character
            else:
                escapedString += character

        return escapedString

File: regexpbuilder.js
class RegExpBuilder {
  constructor() {
    this._flags = '';
    this._literal = [];
    this._groupsUsed = 0;

    this._clear();
  }

  _clear() {
    this._min = -1;
    this._max = -1;
    this._of = '';
    this._ofAny = false;
    this._ofGroup = -1;
    this._from = '';
    this._notFrom = '';
    this._like = '';
    this._either = '';
    this._reluctant = false;
    this._capture = false;
  }

  _flushState() {
    if (this._of != '' || this._ofAny || this._ofGroup > 0 || this._from != '' || this._notFrom != '' || this._like != '') {
      var captureLiteral = this._capture ? '' : '?:';
      var quantityLiteral = this._getQuantityLiteral();
      var characterLiteral = this._getCharacterLiteral();
      var reluctantLiteral = this._reluctant ? '?' : '';
      this._literal.push('(' + captureLiteral + '(?:' + characterLiteral + ')' + quantityLiteral + reluctantLiteral + ')');
      this._clear();
    }
  }

  _getQuantityLiteral() {
    if (this._min != -1) {
      if (this._max != -1) {
        return '{' + this._min + ',' + this._max + '}';
      }
      return '{' + this._min + ',}';
    }
    return '{0,' + this._max + '}';
  }

  _getCharacterLiteral() {
    if (this._of != '') {
      return this._of;
    }
    if (this._ofAny) {
      return '.';
    }
    if (this._ofGroup > 0) {
      return '\\' + this._ofGroup;
    }
    if (this._from != '') {
      return '[' + this._from + ']';
    }
    if (this._notFrom != '') {
      return '[^' + this._notFrom + ']';
    }
    if (this._like != '') {
      return this._like;
    }
  }

  getLiteral() {
    this._flushState();
    return this._literal.join('');
  }
    
  _combineGroupNumberingAndGetLiteral(r) {
    var literal = this._incrementGroupNumbering(r.getLiteral(), this._groupsUsed);
    this._groupsUsed += r._groupsUsed;
    return literal;
  }

  _incrementGroupNumbering(literal, increment) {
    if (increment > 0) {
      literal = literal.replace(/[^\\]\\\d+/g, function (groupReference) {
        var groupNumber = parseInt(groupReference.substring(2)) + increment;
        return groupReference.substring(0, 2) + groupNumber;
      });
    }
    return literal;
  }

  getRegExp() {
    this._flushState();

    return new RegExp(this._literal.join(''), this._flags);
  }

  _addFlag(flag) {
    if (this._flags.indexOf(flag) == -1) {
      this._flags += flag;
    }
    return this;
  }

  ignoreCase() {
    return this._addFlag('i');
  }
  
  multiLine() {
    return this._addFlag('m');
  }
  
  globalMatch() {
    return this._addFlag('g');
  }
  
  startOfInput() {
    this._literal.push('(?:^)');
    return this;
  }
    
  startOfLine() {
    this.multiLine();
    return this.startOfInput();
  }
  
  endOfInput() {
    this._flushState();
    this._literal.push('(?:$)');
    return this;
  }
  
  endOfLine() {
    this.multiLine();
    return this.endOfInput();
  }

  either(r) {
    if (typeof r == 'string') {
      return this._eitherLike(new RegExpBuilder().exactly(1).of(r));
    }
    else {
      return this._eitherLike(r);
    }
  }
  
  _eitherLike(r) {
    this._flushState();
    this._either = this._combineGroupNumberingAndGetLiteral(r);
    return this;
  }
    
  or(r) {
    if (typeof r == 'string') {
      return this._orLike(new RegExpBuilder().exactly(1).of(r));
    }
    else {
      return this._orLike(r);
    }
  }
  
  _orLike(r) {
    var either = this._either;
    var or = this._combineGroupNumberingAndGetLiteral(r);
    if (either == '') {
      var lastOr = this._literal[this._literal.length - 1];
      lastOr = lastOr.substring(0, lastOr.length - 1);
      this._literal[this._literal.length - 1] = lastOr;
      this._literal.push('|(?:' + or + '))');
    }
    else {
      this._literal.push('(?:(?:' + either + ')|(?:' + or + '))');
    }
    this._clear();
    return this;
  }
    
  neither(r) {
    if (typeof r == 'string') {
      return this.notAhead(new RegExpBuilder().exactly(1).of(r));
    }
    return this.notAhead(r);
  }
  
  nor(r) {
    if (this._min == 0 && this._ofAny) {
      this._min = -1;
      this._ofAny = false;
    }
    this.neither(r);
    return this.min(0).ofAny();
  }
  
  exactly(n) {
    this._flushState();
    this._min = n;
    this._max = n;
    return this;
  }
    
  min(n) {
    this._flushState();
    this._min = n;
    return this;
  }
  
  max(n) {
    this._flushState();
    this._max = n;
    return this;
  }
  
  of(s) {
    this._of = this._sanitize(s);
    return this;
  }
    
  ofAny() {
    this._ofAny = true;
    return this;
  }
  
  ofGroup(n) {
    this._ofGroup = n;
    return this;
  }
  
  from(s) {
    this._from = this._sanitize(s.join(''));
    return this;
  }
    
  notFrom(s) {
    this._notFrom = this._sanitize(s.join(''));
    return this;
  }
  
  like(r) {
    this._like = this._combineGroupNumberingAndGetLiteral(r);
    return this;
  }
  
  reluctantly() {
    this._reluctant = true;
    return this;
  }
    
  ahead(r) {
    this._flushState();
    this._literal.push('(?=' + this._combineGroupNumberingAndGetLiteral(r) + ')');
    return this;
  }
  
  notAhead(r) {
    this._flushState();
    this._literal.push('(?!' + this._combineGroupNumberingAndGetLiteral(r) + ')');
    return this;
  }
  
  asGroup() {
    this._capture = true;
    this._groupsUsed++;
    return this;
  }
    
  then(s) {
    return this.exactly(1).of(s);
  }
  
  find(s) {
    return this.then(s);
  }
  
  some(s) {
    return this.min(1).from(s);
  }
  
  maybeSome(s) {
    return this.min(0).from(s);
  }
  
  maybe(s) {
    return this.max(1).of(s);
  }
    
  something() {
    return this.min(1).ofAny();
  }
  
  anything() {
    return this.min(0).ofAny();
  }
  
  anythingBut(s) {
    if (s.length == 1) {
      return this.min(0).notFrom([s]);
    }
    this.notAhead(new RegExpBuilder().exactly(1).of(s));
    return this.min(0).ofAny();
  }
  
  any() {
    return this.exactly(1).ofAny();
  }
    
  lineBreak() {
    this._flushState();
    this._literal.push('(?:\\r\\n|\\r|\\n)');
    return this;
  }
  
  lineBreaks() {
    return this.like(new RegExpBuilder().lineBreak());
  }
  
  whitespace() {
    if (this._min == -1 && this._max == -1) {
      this._flushState();
      this._literal.push('(?:\\s)');
      return this;
    }
    this._like = '\\s';
    return this;
  }
    
  notWhitespace() {
    if (this._min == -1 && this._max == -1) {
      this._flushState();
      this._literal.push('(?:\\S)');
      return this;
    }
    this._like = '\\S';
    return this;
  }
  
  tab() {
    this._flushState();
    this._literal.push('(?:\\t)');
    return this;
  }
  
  tabs() {
    return this.like(new RegExpBuilder().tab());
  }
    
  digit() {
    this._flushState();
    this._literal.push('(?:\\d)');
    return this;
  }
  
  notDigit() {
    this._flushState();
    this._literal.push('(?:\\D)');
    return this;
  }
  
  digits() {
    return this.like(new RegExpBuilder().digit());
  }
    
  notDigits() {
    return this.like(new RegExpBuilder().notDigit());
  }
  
  letter() {
    this.exactly(1);
    this._from = 'A-Za-z';
    return this;
  }
  
  notLetter() {
    this.exactly(1);
    this._notFrom = 'A-Za-z';
    return this;
  }
    
  letters() {
    this._from = 'A-Za-z';
    return this;
  }
  
  notLetters() {
    this._notFrom = 'A-Za-z';
    return this;
  }
  
  lowerCaseLetter() {
    this.exactly(1);
    this._from = 'a-z';
    return this;
  }
    
  lowerCaseLetters() {
    this._from = 'a-z';
    return this;
  }
  
  upperCaseLetter() {
    this.exactly(1);
    this._from = 'A-Z';
    return this;
  }
  
  upperCaseLetters() {
    this._from = 'A-Z';
    return this;
  }
    
  append(r) {
    this.exactly(1);
    this._like = this._combineGroupNumberingAndGetLiteral(r);
    return this;
  }
  
  optional(r) {
    this.max(1);
    this._like = this._combineGroupNumberingAndGetLiteral(r);
    return this;
  }
  
  _sanitize(s) {
    return s.replace(/([.*+?^=!:${}()|\[\]\/\\])/g, '\\$1');
  }
}

exports.ignoreCase = function () {
  return new RegExpBuilder().ignoreCase();
}

exports.multiLine = function () {
  return new RegExpBuilder().multiLine();
}

exports.globalMatch = function () {
  return new RegExpBuilder().globalMatch();
}

exports.startOfInput = function () {
  return new RegExpBuilder().startOfInput();
}

exports.startOfLine = function () {
  return new RegExpBuilder().startOfLine();
}

exports.endOfInput = function () {
  return new RegExpBuilder().endOfInput();
}

exports.endOfLine = function () {
  return new RegExpBuilder().endOfLine();
}

exports.either = function (r) {
  return new RegExpBuilder().either(r);
}

exports.neither = function (r) {
  return new RegExpBuilder().neither(r);
}

exports.exactly = function (n) {
  return new RegExpBuilder().exactly(n);
}

exports.min = function (n) {
  return new RegExpBuilder().min(n);
}

exports.max = function (n) {
  return new RegExpBuilder().max(n);
}

exports.ahead = function (r) {
  return new RegExpBuilder().ahead(r);
}

exports.notAhead = function (r) {
  return new RegExpBuilder().notAhead(r);
}

exports.then = function (s) {
  return new RegExpBuilder().then(s);
}

exports.find = function (s) {
  return new RegExpBuilder().find(s);
}

exports.some = function (s) {
  return new RegExpBuilder().some(s);
}

exports.maybeSome = function (s) {
  return new RegExpBuilder().maybeSome(s);
}

exports.maybe = function (s) {
  return new RegExpBuilder().maybe(s);
}

exports.anything = function () {
  return new RegExpBuilder().anything();
}

exports.anythingBut = function (s) {
  return new RegExpBuilder().anythingBut(s);
}

exports.any = function () {
  return new RegExpBuilder().any();
}

exports.lineBreak = function () {
  return new RegExpBuilder().lineBreak();
}

exports.whitespace = function () {
  return new RegExpBuilder().whitespace();
}

exports.notWhitespace = function () {
  return new RegExpBuilder().notWhitespace();
}

exports.tab = function () {
  return new RegExpBuilder().tab();
}

exports.digit = function () {
  return new RegExpBuilder().digit();
}

exports.notDigit = function () {
  return new RegExpBuilder().notDigit();
}

exports.letter = function () {
  return new RegExpBuilder().letter();
}

exports.notLetter = function () {
  return new RegExpBuilder().notLetter();
}

exports.lowerCaseLetter = function () {
  return new RegExpBuilder().lowerCaseLetter();
}

exports.upperCaseLetter = function () {
  return new RegExpBuilder().upperCaseLetter();
}

exports.append = function (r) {
  return new RegExpBuilder().append(r);
}

exports.optional = function (r) {
  return new RegExpBuilder().optional(r);
}